contrib/transaction_profiling_analyzer/transaction_profiling_analyzer.py

```diff
@@ -508,155 +514,199 @@ def has_dateparser():
-                add_boundary(this_range_start_key, end_key, count_this_range)
-                count_this_range = 0
+
+        for (start_key, (start_count, end_count)) in self.reads.items():
+            open_count -= end_count
+
+            if opened_this_range >= range_size:
+                add_boundary(this_range_start_key, start_key, opened_this_range, count_this_range)
+                count_this_range = open_count
+                opened_this_range = 0
                 this_range_start_key = None
+
+            count_this_range += start_count
+            opened_this_range += start_count
+            open_count += start_count
+
+            if count_this_range > 0 and this_range_start_key is None:
+                this_range_start_key = start_key
+
+            if end_count > 0:
+                last_end = start_key
+
+        if last_end is None:
+            last_end = b'\xff'
         if count_this_range > 0:
-            add_boundary(this_range_start_key, end_key, count_this_range)
+            add_boundary(this_range_start_key, last_end, opened_this_range, count_this_range)
 
+        wait_for_shard_addresses(output_range_counts, shard_finder, 0, 5)
         return output_range_counts
 
 
 class ShardFinder(object):
-    def __init__(self, db):
+    def __init__(self, db, exclude_ports):
         self.db = db
+        self.exclude_ports = exclude_ports
 
-    @staticmethod
-    @fdb.transactional
-    def _get_boundary_keys(tr, begin, end):
-        tr.options.set_read_lock_aware()
-        return fdb.locality.get_boundary_keys(tr, begin, end)
+        self.tr = db.create_transaction()
+        self.refresh_tr()
+
+        self.outstanding = []
+        self.boundary_keys = list(fdb.locality.get_boundary_keys(db, b'', b'\xff\xff'))
+        self.shard_cache = {}
+
+    def _get_boundary_keys(self, begin, end):
+        start_pos = max(0, bisect_right(self.boundary_keys, begin)-1)
+        end_pos = max(0, bisect_right(self.boundary_keys, end)-1)
+
+        return self.boundary_keys[start_pos:end_pos]
+
+    def refresh_tr(self):
+        self.tr.options.set_read_lock_aware()
+        if not self.exclude_ports:
+            self.tr.options.set_include_port_in_address()
 
     @staticmethod
     @fdb.transactional
     def _get_addresses_for_key(tr, key):
-        tr.options.set_read_lock_aware()
         return fdb.locality.get_addresses_for_key(tr, key)
 
     def get_shard_count(self, start_key, end_key):
-        return len(list(self._get_boundary_keys(self.db, start_key, end_key))) + 1
+        return len(self._get_boundary_keys(start_key, end_key)) + 1
 
     def get_addresses_for_key(self, key):
-        return [a.decode('ascii') for a in self._get_addresses_for_key(self.db, key).wait()]
+        shard = self.boundary_keys[max(0, bisect_right(self.boundary_keys, key)-1)]
+        do_load = False
+        if not shard in self.shard_cache:
+            do_load = True
+        elif self.shard_cache[shard].is_ready():
```

Comment:
It would be cleaner, but yes, it's quite likely we don't need to actually read all of them. I'm not sure how more expensive it would be to fetch all of them if we only needed a fraction.

Reply:
Okay